docs/machine-learning/how-to-guides/inspect-intermediate-data-ml-net.md

```diff
@@ -1,75 +1,180 @@
-    [LoadColumn(3)]
-    public string MaritalStatus { get; set; }
+    // Define values for columns being evaluated
+    float size = default;
+    VBuffer<float> historicalPrices = default;
+    float currentPrice = default;
+
+    // Iterate over each row in IDataView
+    while (cursor.MoveNext())
+    {
+        cursor.GetGetter<float>(columns[0]).Invoke(ref size);
+        cursor.GetGetter<VBuffer<float>>(columns[1]).Invoke(ref historicalPrices);
+        cursor.GetGetter<float>(columns[2]).Invoke(ref currentPrice);
+    }
 }
 ```
 
+> [!IMPORTANT]
+> For performance purposes, vectors in ML.NET use [`VBuffer`](xref:Microsoft.ML.Data.VBuffer`1) instead of native collection types (i.e. `Vector`,`float[]`). 
+
+## Preview result of pre-processing or training on a subset of the data
+
+> [!WARNING]
+> Do not use `Preview` in production code because it is intended for debugging and may reduce performance.
+
+The model building process is experimental and iterative. To preview what data would look like after pre-processing or training a machine learning model on a subset of the data, use the [`Preview`](xref:Microsoft.ML.DebuggerExtensions.Preview*) method which returns a [`DataDebuggerPreview`](xref:Microsoft.ML.Data.DataDebuggerPreview). The result is an object with `ColumnView` and `RowView` properties which are both an [`IEnumerable`](xref:System.Collections.Generic.IEnumerable`1) and contain the values in a particular column or row. Specify the number of rows to apply the transformation to with the `maxRows` parameter.
+
+![Data Debugger Preview Object](./media/inspect-intermediate-data-ml-net/data-debugger-preview-01.png)
+
+### Preview contents of an IDataView
+
+The snippet below takes an [`IDataView`](xref:Microsoft.ML.IDataView) inspects the contents of the rows and columns.
+
 ```csharp
-// Create a new context for ML.NET operations. It can be used for exception tracking and logging,
-// as a catalog of available operations and as the source of randomness.
-var mlContext = new MLContext();
-
-// Read the data into a data view.
-var data = mlContext.Data.ReadFromTextFile<InspectedRow>(dataPath, hasHeader: true);
-
-// Start creating our processing pipeline. For now, let's just concatenate all the text columns
-// together into one.
-var pipeline = mlContext.Transforms.Concatenate("AllFeatures", "WorkClass", "Education", "MaritalStatus");
-
-// Fit our data pipeline and transform data with it.
-var transformedData = pipeline.Fit(data).Transform(data);
-
-// Extract the 'AllFeatures' column.
-// This will give the entire dataset: make sure to only take several row
-// in case the dataset is huge. The is similar to the static API, except
-// you have to specify the column name and type.
-var featureColumns =
-    transformedData
-        .GetColumn<string[]>(mlContext, "AllFeatures")
-        .Take(20)
-        .ToArray();
+// Preview IDataView
+DataDebuggerPreview dataPreview = data.Preview(maxRows:5);
```

Comment:
Preview() should only be used inside debugger. we should not write code snippets with it, since Preview() should never be used in code. It's only a tool for using with debugger.

Reply:
I added a warning for it? Do you suggest completely removing this section?

![image](https://user-images.githubusercontent.com/46974588/56929026-92b59580-6aa6-11e9-86c2-94aa775053ad.png)

Reply:
let's not use code snippets for Preview() at all. only show it being used from VisualStudio or some other debugger watch window, with pictures to show how it looks like in the debugger window. If you write it in code, please will use it. No amount of warning can stop that :)

FYI, we have removed Preview() from all our API reference samples, for the same reason.

---
In reply to: [279552086](https://github.com/dotnet/docs/pull/11813#discussion_r279552086) [](ancestors = 279552086)

Reply:
#resolved